### ISSUE TITLE
Limit the number of tasks to 1 for a default install of the BOINC client

### ISSUE THREAD

**Issue by lfield** [BOINC/boinc#1569]
By default the BOINC client will start one task per core. If this happens for our application, especially if there is 2GB per core or less, the machine can freeze. This is not the best first impression to give newcomers and so must be avoided. The decision on the amount of resources that a host can use for BOINC tasks and hence number of parallel tasks that the host can run, is a client side responsibility. If the client would like another task it will ask the server which may or may not assign one.

We  would like to ensure that by default only one task will be assigned and the volunteer has to perform a specific conscious action to increase this limit. Ideally this would be something that could be configurable by the BOINC GUI and the default would be 1 task.

**Comment by AenBleidd** [comment]
Hello @lfield,

I think total memory limitation is better in this case because user can control the amount of memory he wants to be used by non-BOINC applications.

Best regards,
Vitalii

**Comment by ChristianBeer** [comment]
There are already two options that control the number of in progress tasks per application/project.
One is a project setting: http://boinc.berkeley.edu/trac/wiki/ProjectOptions#Joblimitsadvanced
The other is a Client setting: http://boinc.berkeley.edu/wiki/client_configuration#Application_configuration
You would need to test which one has precedence if both are set. Developing a GUI for the Manager to manage an app_config.xml seems nice.

**Comment by lfield** [comment]
After investigating the options, the memory limits do not work well for VM applications. The project settings will affect everyone and they can not be overridden by the volunteer. The client-side settings are good for the experienced volunteer but we need to place default values there for the novice. I think that if we could provide a default app_config.xml, this would solve our problem. We tried doing this by distributing it as a file for our app but it will overwrite changes made when we release a new application version. 

The fact that a GUI does exist, it should be possible to manage this file via the GUI.

**Comment by GITNE** [comment]
@lfield If I understand your concerns correctly then unfortunately you seem to have come to wrong conclusions based on wrong assumptions. You assume that the way your application shares system resources with other BOINC applications and the user's other daily applications is perfect, or at least correct. If your app strictly requires at least 2GB or more RAM per task to run smoothly then it is a requirement that is imposed by *you*. So, it is *you* who has to make sure that all requirements you have are met on the target machine either before launching your app or while the app is running. *Your* app has to keep track of available system resources for as long as it is running in order to not interfere with other BOINC apps, user's daily apps, and overall system stability. And, *your* app (just like any other BOINC app) should stop computing or stop using large amounts of memory if resources get exhausted, regardless of what the user may or may not have setup BOINC to do. Keep in mind that distributed volunteer computing is about dynamically harvesting free system resources, where the amount of available resources changes dramatically in a timely manner. It is particularly **NOT** to be confused with dedicated HPC cluster computing where you have resource guarantees. Most of the machines and system resources which are donated to BOINC must also be **shared** with other applications and users. So, developing an app for volunteer computing is like driving on public roads: Your app must *drive* defensively on the target machine. No data or settings can be taken for granted, except for system data (and even this may not be true if running in a VM).

In the end, it does not matter if the default setting is 1 task per CPU or 1 task per core. Why? Because any badly written, resource greedy BOINC app can easily eat up all resources and bring a machine to a grinding halt.

**Comment by romw** [comment]
@GITNE Laurence is an admin for the various projects hosted at CERN (LHC@Home, vLHC@Home, Atlas@home, etc.), in most of those projects they use VirtualBox which provides some pretty strong limitations as to what the maximum amount of RAM/Disk/CPU an application can use.  Their whole environment is virtualized.

In this case when rsc_memory_bound matches what is specified in vbox_job.xml, there is a good chance that everything will stay within the defined memory range.  Unless VirtualBox itself has a memory leak.

As to the larger point, BOINC in fact does keep track of how much resources an application is using and can/will take corrective actions if an application exceeds those limits.

See:
http://boinc.berkeley.edu/trac/wiki/JobTemplates

Specifically the rsc_* elements.

**Comment by lfield** [comment]
@GITNE My background is HTC mainly in the context of Grid computing for the LHC so I am familiar with many of the concepts involved here, but not necessary the internal details of BOINC. As Rom pointed out, "our" application is the vboxwrapper. Our use case is that we would like to provide a safety catch for novice volunteers, a throttle that will only be opened once the volunteer is ready e.g understand computing preferences and have run successful tasks. I would have thought that this is generic feature than other projects may be interested in. More details on this discion can be found in this [ thread](http://lhcathome2.cern.ch/vLHCathome/forum_thread.php?id=1864). 

We are also considering how to configure multi-core VMs, which may use the same mechanism.

**Comment by GITNE** [comment]
Let me see if I understand this correctly; So, projects like LHC@Home or ATLAS@Home etc deploy VirualBox software via BOINC to target machines and then download work units in form of VirtualBox snapshots/instances, right? And, they do that in order to have compartmentalization, to package runtime version dependencies, and to limit the amount of resource pressure put on target machines?

If so then I start having the weirdest thoughts on that. But, before I share my thoughts I would like you to confirm or deny my understanding first.

> As to the larger point, BOINC in fact does keep track of how much resources an application is using and can/will take corrective actions if an application exceeds those limits.

@romw
Yeah, sort of. These corrective measures are only in response after a fact. These are not hard limits or *rails* like the kernel can impose. Lets put it this way; BOINC can monitor misbehavior with limitations and act on this. And yes, it is true, it is different with VMs which indeed have hard limits but this is a totally different story.

**Comment by lfield** [comment]
@GITNE Virtualization is used as it is the only practical way to run the Linux-based HEP applications on Windows. BOINC can manage the resource IF the client has been configured correctly.  For newcomers this is not always the case and so we would like to provide a safety catch which they can release once they are more experienced and understand how to configure the client.

**Comment by TheRealYeti** [comment]
Sorry, that I have to tell you the following but it is my long time experience with BOINC.

First: I'm a user of BOINC since first days, have run a lot of projects and in the last 2 years focus on CERN, especially Atlas@Home.

In the past, with x32 applications, memory was never a problem.

With x64 applications that need a huge amount of RAM BOINC is not really good in dealing with memory. I'm running boxes with 16 GB and 32 GB memory.

BOINC doesn't take care of memory; I have seen several times that BOINC tried to start more Atlas-Apps as the memory would support. This results in "Postponed ...."

I see BOINC switching through local existing tasks, starting new tasks of a project that has already started and paused tasks. 

7.6.33 seems to ignore the setting of "Leave non-GPU taks in memory while suspended"

With the x64 apps and large memory needs it is not possible to let BOINC manage the jobs by its own.

BOINC tends to switch through the available work and parking the paused Apps in memory. This leads to a memory-overload

I have seen this with BOINC 7.6.22 as with earlier versions as with 7.6.33

This is the main reason why I focus more on Atlas as it is impossible to run Atlas-x64 and vLHC-x64 side by side on the same client.

**Comment by GITNE** [comment]
@lfield I am still not convinced that your problem are BOINC's allegedly wrong default settings. Let me clarify to you that you are actually thinking backwards here. You have a very special case. Your initial problem is that you have a Linux‑only app, despite that fact, your users (or whoever) urged you to run it on Windows (presumably because this is numerically the largest platform), and because of this you employed this crude workaround by packaging a virtualization solution with BOINC (note what is packaged *with* what). This workaround may work technically for you but it is still a crude workaround. Your primary problem is that you do not have a **Windows** **port** of your app for the majority of your volunteers. Again, your issue is not about allegedly wrong BOINC default settings. There are good reasons why ports of BOINC for different platforms exist. The BOINC application model never assumed to or was ever intended to manage VM instances on target machines. I am not saying this cannot be done but this is simply outside of BOINC's primary use case scenario.

So, what can you do?
 1. Do not manage other people's VMs.
If your volunteers want to run a Linux‑only application on their Windows machine then **they** should install and manage a virtualization solution on their machines, not you. Besides, it is impossible to do it in an automated way across all possible hardware and software configurations out there in the wild. Volunteers should install the BOINC client on their (Linux) VMs and attach to your project that would download and run your Linux‑only app. In other words, do not try to solve other people's problems which only they can solve.

 2. Port your app to Windows.
Although this may be quite elaborate, we have never had better tools to do that. This is something you should truly consider, especially given the amount of jobs (or CPU time) that the Windows port would potentially crunch.

Besides, I have found this utterly wrong statement about BOINC on [Virtual LHC@home](http://lhcathome2.cern.ch/vLHCathome)'s website.

> This is a project that utilizes the CERN-developed CernVM virtual machine and the BOINC virtualization layer to harness volunteer cloud computing power for full-fledged LHC event physics simulation on volunteer computers.

BOINC is **NOT** a virtualization layer! It is a distributed application for code download, job scheduling, and result retrieval. BOINC has **NOTHING** to do with virtualization. If you think of BOINC like in the above statement then no wonder your assumptions and also requests are flawed.
Furthermore, cloud computing specifically does not include volunteer computing. In cloud computing, by definition remote computing resources are **managed** by a **professional** entity. Again, volunteer computing has **NOTHING** to do with *the* *cloud* or with **professional** **management**. More precisely, cloud computing indeed does come with resource and availability guarantees. Volunteer computing does not provide any guarantees. Man, you really need to get your head around concepts before you act. This will keep you from producing crude workarounds and also from complaining to other projects that they should adjust to your crude workarounds. Crazy! :laughing:

@TheRealYeti Nobody said that BOINC is perfect. Certainly, there is still room for improvement, especially in the area of memory monitoring and management. However, this is an entirely different topic and not connected to the request made by @lfield here.

**Comment by romw** [comment]
@GITNE Generally speaking, it isn't practical to have CERN port their whole software stack to Windows.  CERNVMFS, which the CERNVM uses internally exposes roughly 50TB worth of software to the VM.  Admittedly that is a collection of several "versions" of the software stack which each "version"/snapshot being roughly 5TB is size if I recall correctly.  Once or twice a year they generate a new snapshot of the software stack for general use throughout the organization.

BOINC's primary mission is to connect volunteers to researchers.  Provide a mechanism to engage the public in scientific research.  We have a lot of latitude in how we accomplish that mission.

CERN's use of BOINC is a good thing and their exploration of BOINC+VirtualBox as a means of supporting large scale scientific research is even a better thing.

The trick in all of this is determining the right set of policies to be successful and that all depends on how well we can make virtualized environments behave on Volunteer's machines.

**Comment by lfield** [comment]
To bring this back on topic, BOINC already provides a number of ways to limit tasks e,g in the config_aux.xml and app_config.xml. This request is to provide a default limit that a volunteer can override. Constructive suggestions on how to achieve this, if possible in a way that would be generic for all projects, are welcome. The consensus is to have a default value in the project preferences that is considered by the scheduler. We have already implemented the part in the project preferences and are now looking at modifying the scheduler in the sched_limits.h. Is it just as simple as adding an additional if statement [here](https://github.com/Ashod/Boinc/blob/master/sched/sched_limit.h#L140). I am guessing that a function already exists to do the properties lookup for he user.

**Comment by ChristianBeer** [comment]
I still don't fully understand what the real problem is.
If the problem is over-committing the resources (memory in this case) of a host with VM tasks than there is a problem on the Client side in which it does not keep track of available memory. This will resurface as soon as people deactivate the limit that is imposed by default. If the Client starts a job that needs 2GB per Core but those are not available then the Client is to blame not the project. This assumes of course that the project set the rsc_memory_bound accordingly.
If the problem is that the client downloads extra tasks and builds up a queue with unstarted tasks than limiting the number of in progress tasks on the server side makes sense. But since the server does not know what's happening on the client you can only limit to `in_progress == cores` and let the Client decide how much concurrent VMs it can start at a given time with the given memory restrictions.

**Comment by TheRealYeti** [comment]
> and let the client decide how much concurrrent VMs it can start at a given time with the given memory 

> 

This is exact the problem. The client first makes the right decisions and starts a realistic number of tasks. But then, when project switch-time happens, the client suspends these VMs, keeps the WU in memory and starts WUs from another project. And if this is also a VM-Project with memory-intensive WUs, this leads to "postponed" or even crashed VMs.

**Comment by lfield** [comment]
Overcommitment occurs for reasons we know, reasons we don't know and reasons that are unforeseen. This request is for a safety feature that is trivial to implement and prevents first time volunteers from having potential problems with their initial attempts. We can follow up the known reasons with other issues and keep investigating to uncover the others. In the meantime and for the unforeseen reasons, this feature will be extremely beneficial.

**Comment by lfield** [comment]
I am closing this issue as there has been no progress.
